### PR TITLE
Added support for geolocalized search engines

### LINKED ISSUE #248: Verify that FxR sends correct search code to Google when user searches from URL bar
Hardware: ALL
Steps to Reproduce: NA
Actual Behavior: NA
Expected Behavior: Verify that FxR sends correct search code to Google (as the sole/default search engine in v1.0) such that URL bar searches are sent to Google appropriately and Mozilla gets credit/paid.
Recommendation: Same

### PR DESCRIPTION
**Pull request by keianhzo**
Fixes #248 Added support for custom search engines based on geolocalization data

### CODE REVIEW

**Comment by bluemarvin** [comment]
Not sure how to validate this PR.

**Comment by keianhzo** [comment]
Good question. What I did during testing is just forcing the country code at `Engine.getEngine`. Maybe there is some way of tricking the Mozilla Geolocation API to return a specific country data @cvan ? Or using a proxy directly set in the WiFi adapter?

File: app/src/main/res/values/non_I10n.xml [hunk 1]
```diff
@@ -5,6 +5,23 @@
...
+    <!-- Google -->
+    <string name="google_params">q=%1$s&amp;ie=utf-8&amp;oe=utf-8&amp;keyword=firefox-b-ab&amp;searchbar=firefox-b</string>
+    <string name="google_us_params">q=%1$s&amp;ie=utf-8&amp;oe=utf-8&amp;keyword=firefox-b-1-ab&amp;searchbar=firefox-b-1</string>
+    <string name="google">https://www.google.com/search</string>
+    <!-- Baidu -->
+    <string name="baidu_params">wd=%1$s&amp;tn=monline_dg&amp;ie=utf-8</string>
+    <string name="baidu_cn">https://www.baidu.com/baidu</string>
+    <!-- Yandex -->
+    <string name="yandex_params">searchbar=2186618&amp;keyword=2186621&amp;contextmenu=2186623&amp;homepage=2186617&amp;newtab=2186620&amp;text=%1$s</string>
+    <string name="yandex_ru">https://yandex.ru/search</string>
```
Comment (cvan) [inline app/src/main/res/values/non_I10n.xml]: what about `AZ` and `KK` for Yandex? (per https://dxr.mozilla.org/mozilla-central/source/browser/components/search/searchplugins and https://location-leaderboard.services.mozilla.com/api/v1/locations/countries/)

Comment (cvan) [inline app/src/main/res/values/non_I10n.xml]: also, I believe Japan (`JP`) uses Yahoo!: https://dxr.mozilla.org/mozilla-central/source/browser/components/search/searchplugins/yahoo-jp.xml

Comment (cvan) [reply app/src/main/res/values/non_I10n.xml]: also, Slovakia (`SK`) seems to have several search engines, but I'm not sure which is the default one: https://dxr.mozilla.org/mozilla-central/source/browser/components/search/searchplugins

Comment (cvan) [reply app/src/main/res/values/non_I10n.xml]: this doesn't look up to date (notice: Yahoo! default for `en-US`), but I found this: https://github.com/mozilla-services/absearch/blob/master/data/config.json (https://search.services.mozilla.com/)

Comment (cvan) [reply app/src/main/res/values/non_I10n.xml]: also, see https://github.com/mozilla-mobile/focus-android/issues/2374 and https://dxr.mozilla.org/mozilla-central/source/mobile/android/base/java/org/mozilla/gecko/search/SearchEngineManager.java

File: app/src/common/shared/org/mozilla/vrbrowser/search/GeolocationTask.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            GeolocationData data = new GeolocationData(null, null);
+
+            JSONObject json;
+            try {
+                json = new JSONObject(aGeolocationJson);
+                data.mCountryCode = (String) json.get("country_code");
+                data.mCountryName = (String) json.get("country_name");
+
+            } catch (JSONException e) {
+                Log.e(LOGTAG, "Error parsing geolocation data: " + e.getLocalizedMessage());
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/search/GeolocationTask.java]: perhaps if `json.get("error")` exists, log that error (per the [API docs](https://mozilla.github.io/ichnaea/api/region.html#response))?

File: app/src/common/shared/org/mozilla/vrbrowser/search/GeolocationTask.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        }
+
+        public static GeolocationData parse(String aGeolocationJson) {
+            GeolocationData data = new GeolocationData(null, null);
+
+            JSONObject json;
+            try {
+                json = new JSONObject(aGeolocationJson);
+                data.mCountryCode = (String) json.get("country_code");
+                data.mCountryName = (String) json.get("country_name");
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/search/GeolocationTask.java]: we should probably handle `json.get("fallback")` if it exists too (per the [API code](https://github.com/mozilla/ichnaea/blob/36e395a/ichnaea/api/locate/views.py#L123-L124))

**Comment by cvan** [comment]
@keianhzo I looked at the source code for https://location.services.mozilla.com/v1/country, and I found [tests for the endpoint which simply mock the response](https://github.com/mozilla/ichnaea/blob/284f795/ichnaea/api/locate/tests/test_region_v1.py#L16-L44). if we can do something in the tests, that'd be ideal. otherwise, for manually testing, using a tool such as [Charles](https://www.charlesproxy.com/) to proxy the requests is probably easiest. or allowing a local override for the [const in the XML](https://github.com/MozillaReality/FirefoxReality/blob/6d153f04/app/src/main/res/values/non_I10n.xml#L13) for dev/test environments could work.

**Comment by keianhzo** [comment]
Thanks a lot for the review @cvan ! Agree with all the suggestions, just one question about the new country codes you mentiones. At https://github.com/MozillaReality/FirefoxReality/issues/248#issuecomment-412278211 @larsbergstrom specified a fixed set of countries, should I add the new ones you suggest (AZ, SK, JP)?

**Comment by cvan** [comment]
@keianhzo: you're welcome. it looks like we can ignore [`fallback`](https://github.com/mozilla/ichnaea/blob/36e395a/ichnaea/api/locate/views.py#L25) ([test](https://github.com/mozilla/ichnaea/blob/284f795/ichnaea/api/locate/tests/test_region_v1.py#L117)), since that's used only if we pass `fallbacks` in the request body. you can ignore my comment about `fallback`.

as for the other items, we should handle and log the `error` object if that appears in the response, so we can keep track of whether particular users' requests are getting ratelimited, blacklisted, etc.

and as for the regions to enable, I'll defer that to @larsbergstrom.

thanks!

File: app/src/common/shared/org/mozilla/vrbrowser/search/GeolocationTask.java (hunk 3 above)
Comment (cvan) [reply app/src/common/shared/org/mozilla/vrbrowser/search/GeolocationTask.java]: per my reply elsewhere, ignore this

**Comment by larsbergstrom** [comment]
@mkaply Can you please answer a few questions?
1) Should Japan (JP) use Yahoo!: https://dxr.mozilla.org/mozilla-central/source/browser/components/search/searchplugins/yahoo-jp.xml
2) Should AZ and KK use Yandex?
3) Slovakia (SK) seems to have several search engines, but I'm not sure which is the default one: https://dxr.mozilla.org/mozilla-central/source/browser/components/search/searchplugins

Thanks!

**Comment by keianhzo** [comment]
Fixed @cvan review suggestions

File: app/src/common/shared/org/mozilla/vrbrowser/search/GeolocationTask.java [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+            return new GeolocationError(aCode, aMessage);
+        }
+
+        public static GeolocationError parse(String aGeolocationJson) {
+            GeolocationError data = null;
+
+            JSONObject json;
+            try {
+                json = new JSONObject(aGeolocationJson);
+                JSONObject error = json.getJSONObject("error");
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/search/GeolocationTask.java]: :+1:

File: app/src/common/shared/org/mozilla/vrbrowser/search/GeolocationTask.java [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        if (mDelegate != null)
+            mDelegate.onGeolocationRequestStarted();
+
+        GeolocationTaskResponse result;
+        do {
+            result = executeGeoLocationRequest();
+
+            if (result.responseType == ResponseType.ERROR) {
+                if (mRetryCount <= mRetries - 1) {
+                    Log.e(LOGTAG, "Geolocation request error, retrying... " + (mRetryCount + 1));
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/search/GeolocationTask.java]: maybe print this?

```
"attempt #" + (mRetryCount + 1)
```

File: app/src/common/shared/org/mozilla/vrbrowser/search/SearchEngine.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            Engine engine = new Engine();
+            engine.mUrlResource = urlResource;
+            engine.mQueryResource = queryResource;
+
+            return engine;
+        }
+
+        @NonNull
+        public static Engine getEngine(@NonNull GeolocationTask.GeolocationData data) {
+            String countryCode = data.getCountryCode().toUpperCase();
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/search/SearchEngine.java]: thanks :+1:

File: app/src/common/shared/org/mozilla/vrbrowser/search/SearchEngine.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        return mEngine.getSearchQuery(mContext, aQuery);
+    }
+
+    public void update() {
+        if (!isUpdating) {
+            mTask = new GeolocationTask(mEndpoint, this);
+            mTask.execute();
+
+        } else {
+            Log.i(LOGTAG, "Geolocation update cancelled, previous update already running");
```
Comment (cvan) [inline app/src/common/shared/org/mozilla/vrbrowser/search/SearchEngine.java]: nice 👍

File: app/src/main/res/values/non_I10n.xml [hunk 8]
```diff
@@ -5,6 +5,24 @@
...
+
+    <!-- SEARCH ENGINES -->
+    <string name="geolocation_api_url">https://location.services.mozilla.com/v1/country</string>
+    <!-- Google -->
+    <string name="search_google_params">q=%1$s&amp;ie=utf-8&amp;oe=utf-8&amp;keyword=firefox-b-ab&amp;searchbar=firefox-b</string>
+    <string name="search_google_us_params">q=%1$s&amp;ie=utf-8&amp;oe=utf-8&amp;keyword=firefox-b-1-ab&amp;searchbar=firefox-b-1</string>
+    <string name="search_google">https://www.google.com/search</string>
+    <string name="search_google_us">https://www.google.com/search</string>
+    <!-- Baidu -->
+    <string name="search_baidu_params">wd=%1$s&amp;tn=monline_dg&amp;ie=utf-8</string>
```
Comment (cvan) [inline app/src/main/res/values/non_I10n.xml]: thanks for prefixing these 👍

File: app/src/main/res/values/non_I10n.xml [hunk 9]
```diff
@@ -5,6 +5,24 @@
...
     <string name="private_policy_url">https://www.mozilla.org/privacy/firefox/</string>
     <string name="private_report_url">https://mixedreality.mozilla.org/fxr/report?src=browser-fxr&amp;label=browser-firefox-reality&amp;url=%1$s</string>
+
+    <!-- SEARCH ENGINES -->
+    <string name="geolocation_api_url">https://location.services.mozilla.com/v1/country</string>
+    <!-- Google -->
+    <string name="search_google_params">q=%1$s&amp;ie=utf-8&amp;oe=utf-8&amp;keyword=firefox-b-ab&amp;searchbar=firefox-b</string>
+    <string name="search_google_us_params">q=%1$s&amp;ie=utf-8&amp;oe=utf-8&amp;keyword=firefox-b-1-ab&amp;searchbar=firefox-b-1</string>
+    <string name="search_google">https://www.google.com/search</string>
+    <string name="search_google_us">https://www.google.com/search</string>
```
Comment (cvan) [inline app/src/main/res/values/non_I10n.xml]: thanks for separating these out 👍

File: app/src/main/res/values/non_I10n.xml [hunk 10]
```diff
@@ -5,6 +5,24 @@
     <string name="settings_key_telemetry">settings_telemetry</string>
     <string name="settings_key_session_id">session_id</string>
     <string name="settings_key_previous_session_id">previous_session_id</string>
+    <string name="settings_key_geolocation_data">geolocation_data</string>
     <string name="private_policy_url">https://www.mozilla.org/privacy/firefox/</string>
     <string name="private_report_url">https://mixedreality.mozilla.org/fxr/report?src=browser-fxr&amp;label=browser-firefox-reality&amp;url=%1$s</string>
+
+    <!-- SEARCH ENGINES -->
+    <string name="geolocation_api_url">https://location.services.mozilla.com/v1/country</string>
```
Comment (cvan) [inline app/src/main/res/values/non_I10n.xml]: perfect :+1:

**Pull request merged**